### PR TITLE
Add options for swap discarding (TRIM)

### PR DESCRIPTION
**Pull request by viric**
###### Motivation for this change
swap on ssd will benefit from TRIM/discard so the writes to the ssd go faster with less wearing.

It works in my encrypted swap0.
```
lsblk -D:
sdb                    0      512B       2G         0
├─sdb1                 0      512B       2G         0
├─sdb2                 0      512B       2G         0
├─sdb3                 0      512B       2G         0
│ └─root               0      512B       2G         0
└─sdb4                 0      512B       2G         0
  └─swap0              0      512B       2G         0
```

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by bb2020** [comment]
Isn't this change the same as adding the following configuration? (unless `swap` partition is in a different device).
```
fileSystems."swap".options = [ "discard" ];
boot.initrd.luks.devices."cryptroot".allowDiscards = true;
```

If `TRIM` support is not properly implemented by the SSD vendor, there is a risk of data loss. There should be a
warning if this is merged. Also enabling `TRIM` on an encrypted area is a potential security concern.
This has been explained here very well: http://asalor.blogspot.com/2011/08/trim-dm-crypt-problems.html.

File: nixos/modules/tasks/encrypted-devices.nix [hunk 1]
```diff
@@ -46,6 +46,12 @@ let
           so the keyfile path should usually start with "/mnt-root/".
         '';
       };
+
+      allowDiscards = mkOption {
+        default = false;
+        type = types.bool;
+        description = "cryptsetup argument --allow-discard on open";
```
Comment (symphorien) [inline nixos/modules/tasks/encrypted-devices.nix:53]: this explains how it works but not what it does.

Comment (viric) [reply nixos/modules/tasks/encrypted-devices.nix]: Now I copied the allowDiscards description from luksroot. Thank you!

**Comment by viric** [comment]
> Isn't this change the same as adding following configuration? (unless `swap` partition is in a different device)
> 
> ```
> fileSystems."swap".options = [ "discard" ];
> boot.initrd.luks.devices."cryptroot".allowDiscards = true;
> ```

Only luksroot had allowDiscards, not the general 'encrypted-devices.nix'. I didn't know I could override options for swap that way either. Does that work? Maybe it should say 'swap0'?

I'm using this change this way:
```
  swapDevices = [ {
       device = "/dev/mapper/swap0";
       discard = true;
       encrypted.enable = true;
       encrypted.blkDev = "/dev/disk/by-id/ata-WDC_WDS120G2G0A-00JH30_1806BB803081-part4";
       encrypted.label = "swap0";
       encrypted.allowDiscards = true;
       } ];
```

> If `TRIM` support is not properly implemented by an SSD vendor, there is a risk of data loss. There should be a
> warning if this PR is merged. Also enabling `TRIM` on an encrypted area is a potential security concern.
> This has been explained here very well: http://asalor.blogspot.com/2011/08/trim-dm-crypt-problems.html.

I agree. I copied the documentation from luksroot.nix

**Comment by bb2020** [comment]
> Only luksroot had allowDiscards, not the general 'encrypted-devices.nix'. I didn't know I could override options for swap that way either. Does that work? Maybe it should say 'swap0'?

You should change the names according to your partition naming. I haven't tested it, but it should work. `fsType = "swap"` attribute should be valid for `fileSystems` option.

Is it a good idea to mount `swap` with `discard`? It may stall disk operations. I would prefer mounting `/` with `discard` and running periodic `fstrim` on swap area.

**Comment by viric** [comment]
On Sun, Jan 10, 2021 at 03:26:28PM -0800, bb2020 wrote:
> > Only luksroot had allowDiscards, not the general 'encrypted-devices.nix'. I didn't know I could override options for swap that way either. Does that work? Maybe it should say 'swap0'?
> Is it a good idea to mount `swap` with `discard`? It may stall disk operations. I would prefer mounting `/` with `discard` and running periodic `fstrim` on swap area.

The manpage (swapon) says that the TRIM runs asynchronously, so maybe they
will not stall. In any case, I could add discard options (once, pages,
both), if you want.

-- 
(Escriu-me xifrat si saps PGP / Write ciphered if you know PGP)
PGP key 7CBD1DA5 - https://emailselfdefense.fsf.org/

File: nixos/modules/config/swap.nix [hunk 2]
```diff
@@ -124,6 +124,17 @@ let
...
 
+      discard = mkOption {
+        default = false;
+        type = types.bool;
+        description = ''
+          Use swapon -d, and if randomEncryption, make the encrypted mapping
+          with --allow-discards.
+
+          If you use encrypted.enable for this swap device, you should enable
+          there encrypted.allowDiscards.
```
Comment (symphorien) [inline nixos/modules/config/swap.nix:135]: this could be done automatically, right ?

Comment (viric) [reply nixos/modules/config/swap.nix:135]: hm I'm not sure how. Any suggestion?

**Comment by ajs124** [comment]
How does this relate to #107728?

**Comment by bb2020** [comment]
> How does this relate to #107728?

Oh, that looks like the official way to achieve this.

**Comment by kurnevsky** [comment]
Looks like changes in #107728 will work only with unencrypted swap, so discards still should be enabled in in luks.

**Comment by bb2020** [comment]
It is good to see #107728 is merged.

> Looks like changes in #107728 will work only with unencrypted swap, so discards still should be enabled in in luks.

I suppose you can use LVM to create `swap` and `root` areas on a `cryptroot` LUKS partition and then use this to enable `discard` on both. `boot.initrd.luks.devices."cryptroot".allowDiscards = true;`

TRIM on LUKS should not be enabled by default as there are security implications. It is explained [here](http://asalor.blogspot.com/2011/08/trim-dm-crypt-problems.html). It should be up to the user.

**Comment by kurnevsky** [comment]
Here I added PR for this: https://github.com/NixOS/nixpkgs/pull/129408

> I suppose you can use LVM to create swap and root areas on a cryptroot LUKS partition and then use this to enable discard on both.

I'd prefer not to use LVM as I don't need it and use btrfs subvolumes instead.

> TRIM on LUKS should not be enabled by default as there are security implications. It is explained here. It should be up to the user.

Right, it's disabled by default via `discardPolicy`.

**Comment by Artturin** [comment]
superseded by #107728 and #129408

additionally #128032 will allow people to set their own swap options

**Pull request closed without merging**